src/variational/VariationalInference.jl

```diff
@@ -0,0 +1,171 @@
+module Variational
+
+using ..Core, ..Core.RandomVariables, ..Utilities
+using Distributions, Bijectors
+using ProgressMeter, LinearAlgebra
+using ..Turing: PROGRESS
+using ..Turing: Model, SampleFromPrior, SampleFromUniform
+using ..Turing: Turing
+using Random: AbstractRNG
+
+using ForwardDiff
+using Flux.Tracker, Flux.Optimise
+
+import ..Core: getchunksize, getADtype
+
+export
+    vi,
+    ADVI,
+    ELBO
+
+
+abstract type VariationalInference{AD} end
+
+getchunksize(::T) where {T <: VariationalInference} = getchunksize(T)
+getchunksize(::Type{<:VariationalInference{AD}}) where AD = getchunksize(AD)
+getADtype(alg::VariationalInference) = getADtype(typeof(alg))
+getADtype(::Type{<: VariationalInference{AD}}) where {AD} = AD
+
+abstract type VariationalObjective end
+
+const VariationalPosterior = Distribution{Multivariate, Continuous}
+
+"""
+    rand(vi::VariationalInference, num_samples)
+
+Produces `num_samples` samples for the given VI method using number of samples equal to `num_samples`.
+"""
+function rand(vi::VariationalPosterior, num_samples) end
+
+"""
+    objective(vi::VariationalInference, q::VariationalPosterior, model::Model, args...)
+
+Computes the variational objective to be optimized for a given VI method.
+"""
+function objective(vi::VariationalInference, q::VariationalPosterior, model::Model, num_samples) end
+
+# (::VariationalObjective)(vi::VariationalInference, model::Model, num_samples, args...; kwargs...) = begin
+# end
+
+
+"""
+    grad!(vo, vi::VariationalInference, q::VariationalPosterior, model::Model, θ, out, args...)
+
+Computes the gradients used in `optimize!`. Default implementation is provided for `VariationalInference{AD}` where `AD` is either `ForwardDiffAD` or `TrackerAD`. This implicitly also gives a default implementation of `optimize!`.
+
+Variance reduction techniques, e.g. control variates, should be implemented in this function.
+"""
+grad!(vo, vi::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ, out, args...) where AD =
+    error("Turing.Variational.grad!: unmanaged variational inference algorithm: $(typeof(alg))")
+
+"""
+    vi(model::Model, alg::VariationalInference)
+    vi(model::Model, alg::VariationalInference, q::VariationalPosterior)
+
+Constructs the variational posterior from the `model` and performs the optimization following the configuration of the given `VariationalInference` instance.
+"""
+vi(model::Model, alg::VariationalInference) = 
+    error("Turing.Variational.vi: unmanaged variational inference algorithm: $(typeof(alg))")
+vi(model::Model, alg::VariationalInference, q::VariationalPosterior) =
+    error("Turing.Variational.vi: unmanaged variational inference algorithm: $(typeof(alg))")
+
+# default implementations
+function grad!(vo, alg::VariationalInference{AD}, q::VariationalPosterior, model::Model, θ::AbstractVector{T}, out::DiffResults.MutableDiffResult, args...) where {T <: Real, AD <: ForwardDiffAD}
```

Comment:
Please see the `Method Definitions` section of the [style guide](https://turing.ml/docs/style-guide/) for method definitions spread over multiple lines.